emu_mps/mps_backend_impl.py

```diff
@@ -690,6 +691,106 @@ def fill_results(self) -> None:
+
+    def convergence_check(self, energy_tolerance: float) -> bool:
+        if self.previous_energy is None or self.current_energy is None:
+            return False
+        return abs(self.current_energy - self.previous_energy) < energy_tolerance
+
+    def progress(self) -> None:
+        if self.is_finished():
+            return
+
+        # perform one two-site energy minimization and update
+        assert self.state.orthogonality_center is not None
+        idx = self.state.orthogonality_center
+
+        assert (
+            self.swipe_direction == SwipeDirection.LEFT_TO_RIGHT
+            or self.swipe_direction == SwipeDirection.RIGHT_TO_LEFT
+        )
+        if self.swipe_direction == SwipeDirection.LEFT_TO_RIGHT:
+            left_idx, right_idx = idx, idx + 1
+        elif self.swipe_direction == SwipeDirection.RIGHT_TO_LEFT:
+            left_idx, right_idx = idx - 1, idx
+
+        new_L, new_R, energy = minimize_energy_pair(
+            state_factors=self.state.factors[left_idx : right_idx + 1],
+            ham_factors=self.hamiltonian.factors[left_idx : right_idx + 1],
+            baths=(self.left_baths[-1], self.right_baths[-1]),
+            orth_center_right=(self.swipe_direction == SwipeDirection.LEFT_TO_RIGHT),
+            config=self.config,
+            residual_tolerance=self.residual_tolerance,
+        )
+        self.state.factors[left_idx], self.state.factors[right_idx] = new_L, new_R
+        self.current_energy = energy
+
+        # updating baths and orthogonality center
+        if self.swipe_direction == SwipeDirection.LEFT_TO_RIGHT:
+            self.left_baths.append(
+                new_left_bath(
+                    self.get_current_left_bath(),
+                    self.state.factors[left_idx],
+                    self.hamiltonian.factors[right_idx],
+                ).to(self.state.factors[right_idx].device)
+            )
+            self.right_baths.pop()
+            self.state.orthogonality_center += 1
+
+            if self.state.orthogonality_center == self.qubit_count - 1:
+                self.swipe_direction = SwipeDirection.RIGHT_TO_LEFT
+
+        elif self.swipe_direction == SwipeDirection.RIGHT_TO_LEFT:
+            self.right_baths.append(
+                new_right_bath(
+                    self.get_current_right_bath(),
+                    self.state.factors[right_idx],
+                    self.hamiltonian.factors[right_idx],
+                ).to(self.state.factors[left_idx].device)
+            )
+            self.left_baths.pop()
+            self.state.orthogonality_center -= 1
+
+            if self.state.orthogonality_center == 0:
+                self.swipe_direction = SwipeDirection.LEFT_TO_RIGHT
+                self.sweep_count += 1
+                self.sweep_complete()
+
+    def sweep_complete(self) -> None:
+        # This marks the end of one full sweep: checking convergence
+        if self.convergence_check(self.energy_tolerance):
```

Comment:
This can be written with a single indentation level:

```python
        if self.convergence_check(self.energy_tolerance):
            self.timestep_complete()
        elif self.sweep_count + 1 > self.max_sweeps:
            # not converged: restart a new sweep
            raise RuntimeError(
                f"DMRG did not converge after {self.max_sweeps} sweeps"
                )
        else:
            self.previous_energy = self.current_energy
```